pkg/compiler/workflow_compiler.go

```diff
@@ -230,12 +236,23 @@ func (w workflowBuilder) ValidateWorkflow(fg *flyteWorkflow, errs errors.Compile
 			continue
 		}
 
-		if _, foundUpStream := wf.upstreamNodes[nodeID]; !foundUpStream {
-			wf.AddExecutionEdge(c.StartNodeID, nodeID)
+		// Nodes that do not have a upstream dependencies means they do not rely on the workflow inputs to execute.
+		// This is a rare but possible occurrence, by explicitly adding an execution edge from the start node to these
+		// nodes, we ensure that propeller starts executing the workflow by running all such nodes and then their
+		// downstream dependencies.
+		if topLevelNodes.Has(nodeID) {
```

Comment:
this would have caused all kinds of problems for non input graphs?

Reply:
Notice the change doesn't start adding downstream/upstream dependencies... that was already there... the only addition/change is making sure we only add them for top level nodes....